File: statsig-node/src/__tests__/GetExperimentGroups.test.ts
import * as fs from 'node:fs';
import * as path from 'node:path';

import { Statsig, StatsigOptions } from '../../build/index.js';
import { MockScrapi } from './MockScrapi';

describe('Get Experiment Groups', () => {
  let scrapi: MockScrapi;
  let statsig: Statsig;

  beforeAll(async () => {
    scrapi = await MockScrapi.create();

    const dcs = fs.readFileSync(
      path.join(
        __dirname,
        '../../../statsig-rust/tests/data/eval_proj_dcs.json',
      ),
      'utf8',
    );

    scrapi.mock('/v2/download_config_specs', dcs, {
      status: 200,
      method: 'GET',
    });

    scrapi.mock('/v1/log_event', '{"success": true}', {
      status: 202,
      method: 'POST',
    });

    const specsUrl = scrapi.getUrlForPath('/v2/download_config_specs');
    const logEventUrl = scrapi.getUrlForPath('/v1/log_event');
    const options: StatsigOptions = {
      specsUrl,
      logEventUrl,
    };

    statsig = new Statsig('secret-123', options);
    await statsig.initialize();
  });

  afterAll(async () => {
    await statsig.shutdown();
    scrapi.close();
  });

  it('should return the groups for a known experiment', async () => {
    const result = statsig.getExperimentGroups('test_experiment_no_targeting');

    expect(result.isExperimentActive).toBe(true);

    const groupsByName = Object.fromEntries(
      result.groups.map((group) => [group.groupName, group]),
    );

    // Only the experiment group rules are returned (the layerAssignment rule is excluded).
    expect(Object.keys(groupsByName).sort()).toEqual(['Control', 'Test', 'Test2']);
    expect(groupsByName['Control'].returnValue).toEqual({ value: 'control' });
    expect(groupsByName['Control'].ruleId).toBe('54QJztEPRLXK7ZCvXeY9q4');
    expect(groupsByName['Control'].idType).toBe('userID');
    expect(groupsByName['Test'].returnValue).toEqual({ value: 'test_1' });
    expect(groupsByName['Test2'].returnValue).toEqual({ value: 'test_2' });
  });

  it('should return undefined active state for an unknown experiment', async () => {
    const result = statsig.getExperimentGroups('not_an_experiment');

    expect(result.isExperimentActive).toBeUndefined();
    expect(result.groups).toEqual([]);
  });

  it('should return undefined active state for a dynamic config', async () => {
    const result = statsig.getExperimentGroups(
      'test_max_dynamic_config_size_again',
    );

    expect(result.isExperimentActive).toBeUndefined();
    expect(result.groups).toEqual([]);
  });

  it('should return the groups for an inactive experiment', async () => {
    // test_switchback has isActive: false; groups are still returned along with the flag.
    const result = statsig.getExperimentGroups('test_switchback');

    expect(result.isExperimentActive).toBe(false);

    // Only the experiment group rules are returned (non-group rules are excluded).
    const groupNames = result.groups.map((group) => group.groupName).sort();
    expect(groupNames).toEqual(['Control', 'Test']);
  });
});

File: statsig-elixir/test/statsig/get_experiment_groups_test.exs
defmodule Statsig.GetExperimentGroupsTest do
  use ExUnit.Case, async: false

  alias Statsig.Test.MockScrapi
  alias Statsig.Options

  @eval_proj_path Path.expand("../../../statsig-rust/tests/data/eval_proj_dcs.json", __DIR__)
  @eval_proj_data File.read!(@eval_proj_path)

  setup do
    {:ok, mock} = MockScrapi.start_link(@eval_proj_data)

    options = %Options{
      specs_url: mock.specs_url,
      disable_all_logging: true,
      output_log_level: "warn"
    }

    {:ok, _pid} = Statsig.start_link("secret-key", options)
    Statsig.initialize()

    on_exit(fn ->
      Statsig.shutdown()

      if pid = Process.whereis(Statsig) do
        Process.exit(pid, :normal)
      end

      MockScrapi.stop(mock)
    end)

    :ok
  end

  test "returns the groups for a known experiment" do
    {:ok, result} = Statsig.get_experiment_groups("test_experiment_no_targeting")

    assert result.is_experiment_active == true

    groups_by_name =
      Map.new(result.groups, fn group -> {group.group_name, group} end)

    # Only the experiment group rules are returned (the layerAssignment rule is excluded).
    assert Enum.sort(Map.keys(groups_by_name)) == ["Control", "Test", "Test2"]
    assert groups_by_name["Control"].return_value == %{"value" => "control"}
    assert groups_by_name["Control"].rule_id == "54QJztEPRLXK7ZCvXeY9q4"
    assert groups_by_name["Control"].id_type == "userID"
    assert groups_by_name["Test"].return_value == %{"value" => "test_1"}
    assert groups_by_name["Test2"].return_value == %{"value" => "test_2"}
  end

  test "returns ExperimentGroupsResult and ExperimentGroup structs" do
    {:ok, result} = Statsig.get_experiment_groups("test_experiment_no_targeting")

    assert match?(%Statsig.ExperimentGroupsResult{}, result)
    assert Enum.all?(result.groups, &match?(%Statsig.ExperimentGroup{}, &1))
  end

  test "returns nil active state for an unknown experiment" do
    assert {:ok, %Statsig.ExperimentGroupsResult{is_experiment_active: nil, groups: []}} =
             Statsig.get_experiment_groups("nonexistent_experiment")
  end

  test "returns nil active state for a dynamic config" do
    assert {:ok, %Statsig.ExperimentGroupsResult{is_experiment_active: nil, groups: []}} =
             Statsig.get_experiment_groups("test_max_dynamic_config_size_again")
  end

  test "returns the groups for an inactive experiment" do
    # test_switchback has isActive: false; groups are still returned along with the flag.
    {:ok, result} = Statsig.get_experiment_groups("test_switchback")

    assert result.is_experiment_active == false

    # Only the experiment group rules are returned (non-group rules are excluded).
    assert result.groups |> Enum.map(& &1.group_name) |> Enum.sort() == ["Control", "Test"]
  end
end
